### PR TITLE
Ubuntu/Debian support without breaking OSX

### PR DESCRIPTION
**Pull request by nilportugues**
Hello @taylorotwell and pretty much everyone in the Laravel community.. and PHP Community too!
I have put this quick hack together to make everything work for Ubuntu.

## OS Supported
- It works for Ubuntu 16, and should work for Ubuntu 15 versions too and any Debian using **systemd**.

## Current status: 
- :white_check_mark: Linking and parking work, so do unlinking operations
- :white_check_mark: Add to composer.json a script to download https://ngrok.com for OSX or Linux. 
- :white_check_mark: Fallback for missing PBCOPY working!
- :white_check_mark: Support for Ubuntu 14 and Debian using **upstart** by creating another CompatibilityLinux-like file doing `which upstart` to check if upstart is available.

Awesome work Taylor. Loved it.

### CODE REVIEW

**Comment by nilportugues** [comment]
I figured out a way to do this clean and fast.  Working on a mergable version. Keep you posted :+1:

**Comment by ricardobarantini** [comment]
+1

**Comment by edbizarro** [comment]
+1

**Comment by mroedel** [comment]
+1

**Comment by nathenapse** [comment]
+1

**Comment by nilportugues** [comment]
OK guys, I need to test upstart for Ubuntu 14, and I believe I'm done. 

Then I'll update the branch and resolve its conflicts to prepare for merge.

**Comment by ohnotnow** [comment]
It'd only be a couple of lines change for apt-get -> yum to make this work with RHEL/CentOS 7 (from a quick look).  Would you be open to a patch?

**Comment by nilportugues** [comment]
@ohnotnow obviously yeah :+1:

**Comment by ohnotnow** [comment]
Ok-doke - I'll see what I can do. Getting PHP7 installed is actually more work than the rest of it, but I'll see if the RH SCL can be done without major surgery ;-)

**Comment by nilportugues** [comment]
@ohnotnow remember original package does not install php7 or mariadb/mysql for you. 

To port it to RHEL/CentOS just make sure systemd/upstart, the pbcopy alias and ngrok work just fine.

**Comment by ohnotnow** [comment]
Doh! That'll teach me to read the whole script before I start... ;-)

**Comment by adamwathan** [comment]
FWIW Valet works totally fine on PHP 5.5+ so don't stress about PHP 7 👍

**Comment by nilportugues** [comment]
Upstart working. 

Time to resolve conflicts and merge :+1:

**Comment by ohnotnow** [comment]
*Somewhat ironically*, I can't get any of my centos 7 vagrant boxes to build while I'm at home - so I'll have to wait 'till Tuesday or so when I get back to work.  If anyone else wants to jump in feel free :-)

@adamwathan annoyingly RHEL/CentOS 7 come with php 5.4 (oh yes) unless you add extra repo's. No worse than adding a PPA on deb/ubuntu, but it'll probably need pointing out.

**Comment by nilportugues** [comment]
Ready for review and merge

**Comment by radel** [comment]
Wow, that was fast! thank you

**Comment by nilportugues** [comment]
For anyone willing what is required to do CentOS/RHEL/Fedora support. Changes would be: 

- WHICH_INSTALLER = 'which yum'
- WHICH_INSTALLER_PATH = '/bin/yum'
- WHICH_INSTALLER_ERROR = 'Valet requires yum to be installed on your Linux.';
- DNSMASQ_INSTALL = 'sudo -u %s yum install -y dnsmasq; sudo chkconfig dnsmasq on';

If this gets merged I'll put the effort to make it work for CentOS/RHEL/Fedora.

**Comment by maherelgamil** [comment]
+1

**Comment by huglester** [comment]
Wow this is fast and very welcome. Thank you!!!!

**Comment by itsmehran** [comment]
@taylorotwell, How much time we should wait for this pull request to be merged ? :) 
Also thanks to @nilportugues :+1:

**Comment by vs0uz4** [comment]
+1

**Comment by adriaanzon** [comment]
I think it would be nicer if the script first checks if Dnsmasq is already installed and ready, and if it is, skips the package manager part.
That way it would be easier to make it work on other distributions.

**Comment by nilportugues** [comment]
@adriaanzon the version currenlty provided is osx only. To make it work everywhere the proper version must be provided instead

**Comment by nilportugues** [comment]
@taylorotwell is there anything I can do to help out to merge this sooner than later?

**Comment by tpharaoh** [comment]
That would be great @taylorotwell if merged

**Comment by taylorotwell** [comment]
While I appreciate the effort, Valet has been significantly refactored since this PR was made and the entire thing would likely need to be re-worked. The project now has tests, etc.

**Comment by nilportugues** [comment]
@taylorotwell I can catch up. Not a problem.

**Comment by maherelgamil** [comment]
@nilportugues please catch up :)

**Comment by huglester** [comment]
I think best would be to wait until everything settles. Because now @taylorotwell is pushing updates so often, that you will have to spend too much of the time :(

But too sad Taylor did not at least 'accepted' it, so you at least know that it will get merged

**Comment by tpharaoh** [comment]
@taylorotwell @nilportugues both of you, great work.
I was thinking though... most laravel devs got all the tools preinstalled. But effortless php artisan share  (ala valet) would be awesome in laravel 5.3... to allow clients to see work in progress when its time to get feedback.

File: valet.php [hunk 1]
```diff
@@ -93,6 +95,21 @@
...
+ */
+$app->command('which', function ($output) {
+    require __DIR__.'/drivers/require.php';
+
+    $driver = ValetDriver::assign(getcwd(), basename(getcwd()), '/');
+
+    if ($driver) {
+        $output->writeln('<info>This site is served by ['.get_class($driver).'].</info>');
+    } else {
+        $output->writeln('<fg=red>Valet could not determine which driver to use for this site.</>');
```
Comment (QWp6t) [inline valet.php]: Is `</>` the correct syntax? I'm assuming it isn't.

**Comment by ohnotnow** [comment]
I see this is marked as closed now - is this a "not wanted" feature or just this PR is out of date?  The code seems to have settled a bit so I'd be interested in doing a new PR if it'd be considered?

**Comment by nilportugues** [comment]
@ohnotnow I don't know... current code has "brew" everywhere. So introducing this compatibility change would require chaning brew to "packageManager" to make sense.

Unless there's an official statement wanting to merge a future commit on this using the latest code, I won't put effort into this.

Is up to @taylorotwell.

**Comment by huglester** [comment]
Is is very sad that @taylorotwell is ignoring this pr. We saw a lot of improvents after first launch ov valet. But now things settled and this seems as a good addition

**Comment by barryvdh** [comment]
If he doesn't want to support Windows/Ubuntu, you could just release your own fork for Ubuntu?

**Comment by alfredbez** [comment]
There are working linux forks, e.g. https://github.com/cpriego/valet-ubuntu

**Comment by nilportugues** [comment]
@cpriego hopefully will pick up all my code and make it his :+1:

**Comment by mehranRasulian** [comment]
@nilportugues  Why you don't do that yourself ?

**Comment by tomicakorac** [comment]
Well there is "brew" for Linux, isn't there? [Linuxbrew](http://linuxbrew.sh/). Can we not use that? @nilportugues is right, Valet heavily relies on Homebrew, which is a real shame. It would take some time to adopt it for other environments, but I'm sure it will eventually hapen.

**Comment by cpriego** [comment]
@nilportugues I actually tried to do the "support without breaking OSX" part at first but there are so many differences between the operating systems that it did not make any sense trying to make it work together so I just went the specific OS way as it makes for a cleaner, easier to maintain product.

**Comment by nilportugues** [comment]
@mehranRasulian my time is limited. I'm already involved in local communities for Laravel, Symfony and PHP for Barcelona and I'm an active member in all of them.

**Comment by dccampbell** [comment]
Has anyone attempted to utilize Linuxbrew, as @tomicakorac suggested? I'm a bit baffled by all the forks diverging so much when that seems the obvious match to the Homebrew dependency. I've been digging through Valet's code looking for OS-specific stuff, but everything seems pretty agnostic outside of *nix and *brew. Can anyone speak to where the exact pain points are for Linux compatibility?

cc @cpriego @nilportugues

**Pull request closed without merging**